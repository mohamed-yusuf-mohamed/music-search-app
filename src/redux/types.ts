export interface ResponseData {
  artistName: string,
  kind: string,
  trackName: string, 
  artworkUrl60: string   
}

// TODO: types
export interface SearchBarState {
  loading: boolean,
  error: boolean,
  // status: 'loading' | 'loaded' | 'error',
  parsedInput: string,
  input: string,
  data: ResponseData[],
  fetchCount: number
}

const initialState: SearchBarState = {
  // status: ""
  loading: false,
  error: false,
  parsedInput: "",
  input: "",
  data: [],
  fetchCount: 1
};