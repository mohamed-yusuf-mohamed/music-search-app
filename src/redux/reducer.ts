// import { combineReducers } from 'redux'
import {ResponseData} from "./types"

// COUNTER REDUCER
// const counterReducer = (state = 0, { type }) => {
//   switch (type) {
//     case types.INCREMENT:
//       return state + 1
//     case types.DECREMENT:
//       return state - 1
//     case types.RESET:
//       return 0
//     default:
//       return state
//   }
// }

// INITIAL TIMER STATE
const initialState = {
    loading: false,
    parsedInput: "",
    input: "",
    data: [] as ResponseData[],
    fetchCount: 0,
    error: false
}

// TIMER REDUCER
// TODO: types
// TODO: refactor
const reducer = (state = initialState, action: any) => {
  // ...state.data
  // ...action.payload.data
  // action.payload.data
  
  switch (action.type) {

    
    case "FETCH_STARTED":
      return {
        ...state,
        loading: true
        // fetchCount: state.fetchCount + 1
      }
    case "TEST":
      return {
        ...state,
        fetchCount: state.fetchCount + 1,
        data: action.payload.data
        // fetchCount: state.fetchCount + 1
      }
    case "FETCH_SUCCESS":
      return {
        ...state,
        data: action.payload.data,
        // fetched: action.payload.data,
        loading: false,
        fetchCount: state.fetchCount + 1
      }
    case "FETCH_ERROR":
      return {
        ...state,
        error: true,
        loading: false,
        fetchCount: state.fetchCount - 1
      }
    case "CHANGE_INPUT":
      return {
        ...state,
        input: action.payload,
        parsedInput: action.payload.split(' ').join('+')
      }
    default:
      return state
  }
}

// COMBINED REDUCERS
// const reducers = {
//   counter: counterReducer,
//   timer: timerReducer,
// }

export default reducer

// export default combineReducers(reducers)
