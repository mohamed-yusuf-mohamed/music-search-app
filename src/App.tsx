import React, { useState, useCallback } from 'react';
import { useSelector, useDispatch } from './redux/hooks';
// import { handleInput } from './search-bar/actions';
import { fetchData } from './redux/actions';
import { shallowEqual } from 'react-redux'
import SearchInput from "./components/search-input"
import { Button, InputAdornment, TextField, CircularProgress } from "@material-ui/core"
import {SearchRounded} from '@material-ui/icons';
import Box from '@material-ui/core/Box';
import $ from "jquery"
import styled from "styled-components"
import debounce from "lodash/debounce"
import isEqual from "lodash/isEqual"
import DisplayResults from './components/display-results/display-results';


// TODO: delaunay triangles that react to keystrokes

function App() {
  const dispatch = useDispatch();
  // TODO: use ignite eslint
  // track line legnth
  
  const trackScrolling = useCallback((e) => {
    // TODO: refactor
    const bottom = Math.round($(window).scrollTop() || 0) + Math.round($(window).height() || 0) === Math.round($(document).height() || 0)
    if (bottom) {
      return dispatch(fetchData())
    }
  }, [dispatch]);
  const debTrackScrolling = debounce(trackScrolling, 500)

  React.useEffect(() => {
    $(window).on("wheel", debTrackScrolling)
    return () => {
      $(window).off("wheel", debTrackScrolling)
    }
  }, [debTrackScrolling])

  return (
    <Container id="flex-container">
      <Content id="content">
        <Header id="header">
          <SearchInput />
        </Header>
        <DisplayResults />
      </Content>
    </Container>
  );
}

const Header = styled.div`
  background-color: whitesmoke;
  display: flex;
  align-items: flex-end;
  justify-content: center;
  padding: 10rem 5rem 2rem 5rem;

`

const Container = styled.main`
  display: flex; 
  justify-content: center; 
  flex-direction: row; 
  background-color: grey;
`

const Content = styled.div`
  background-color: pink; 
  flex: 0.5; 
  flex-direction: column; 
  display: flex;
`

export default App;

