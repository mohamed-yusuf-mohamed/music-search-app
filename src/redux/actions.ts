import Api from "../api"
import {Dispatch, State, GetState, Thunk} from "./store"
import { Action } from '@reduxjs/toolkit';
import { AnyAction } from '@reduxjs/toolkit';
import { ThunkDispatch } from "./store";

// TODO: change commit author and name on all commits

const api = new Api()

export const reset = () => ({
  type: "RESET"
})

// TODO: resets on a new input

export const load = () => ({
  type: "LOAD"
})

export const newSearch = () => (dispatch: Dispatch) => {
  // console.log('dispatch', dispatch);
  dispatch(reset())
  return (dispatch as ThunkDispatch)(fetchData())
}

export const fetchData = (): Thunk => async (dispatch: Dispatch, getState: GetState) => {
    const state = getState()
    const {parsedInput, page, data} = state
    const offset = page * 10
    // TODO: validate the offset is working as it should
    const url = `https://itunes.apple.com/search?term=${parsedInput}&offset=${offset}&limit=10`

    dispatch(load())

    try {
      const res = await api.fetch(url)
      return dispatch({
        type: "FETCH",
        payload: {
          data: {
            ...data,
            ...res
          }
        }
      })
    } catch(err) {
      // add your logger here for the error
      return dispatch({
        type: "ERROR",
        payload: {
          error: true,
          loading: false
        }
      })
    }
}

export const handleInput = (e: React.ChangeEvent<HTMLInputElement>): AnyAction => ({ 
  type: "INPUT", 
  payload: e.target.value 
})