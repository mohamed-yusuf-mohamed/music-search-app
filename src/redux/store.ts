import { configureStore, ThunkAction, Action, applyMiddleware } from '@reduxjs/toolkit';
import { combineReducers } from 'redux'
// import counterReducer from '../../DELETEME/counter/counterSlice';
import reducer from "./reducer"
import thunk from "redux-thunk"
import { createStore } from 'redux';

// const reducer = combineReducers({counterReducer, appReducer})
// import rootReducer from './reducers/index';

// Note: this API requires redux@>=3.1.0
// export const store = createStore(reducer, applyMiddleware(thunk));

export const store = createStore(reducer, applyMiddleware(thunk))

// export const store = configureStore({
//   reducer: {
//     counter: counterReducer,
//     app: appReducer
//   }
// });

// export const 

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
