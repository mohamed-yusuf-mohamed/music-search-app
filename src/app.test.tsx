import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { store } from './redux/store';
import App from './app';

describe("app", () => {
  it.only('renders', () => {
    const app = render(
      <Provider store={store}>
        <App />
      </Provider>
    );
  
    expect(app).toMatchSnapshot()
  });
})
