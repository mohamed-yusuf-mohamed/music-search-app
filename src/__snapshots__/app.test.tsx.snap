// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app renders 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <main
        class="sc-iqAclL dFOpGW"
        data-testid="app-container"
      >
        <div
          class="sc-crzoAE eQQNpY"
          id="app-content"
        >
          <div
            class="sc-kEqXSa bYPdLh"
            id="app-header"
          >
            <div
              class="MuiFormControl-root MuiTextField-root sc-gtsrHT gePIXd"
            >
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
                  id="search-input"
                  placeholder="Search for your favourite song, artist or album"
                  type="text"
                  value=""
                />
                <div
                  class="MuiInputAdornment-root MuiInputAdornment-positionStart"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M15.5 14h-.79l-.28-.27c1.2-1.4 1.82-3.31 1.48-5.34-.47-2.78-2.79-5-5.59-5.34-4.23-.52-7.79 3.04-7.27 7.27.34 2.8 2.56 5.12 5.34 5.59 2.03.34 3.94-.28 5.34-1.48l.27.28v.79l4.25 4.25c.41.41 1.08.41 1.49 0 .41-.41.41-1.08 0-1.49L15.5 14zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                    />
                  </svg>
                </div>
                <fieldset
                  aria-hidden="true"
                  class="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
                  style="padding-left: 8px;"
                >
                  <legend
                    class="PrivateNotchedOutline-legend-2"
                    style="width: 0.01px;"
                  >
                    <span>
                      ​
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="sc-dlnjwi ksErfN"
            id="display-results-container"
          >
            <div
              class="sc-hKFxyN iiWwtn"
              id="display-results-content"
            />
          </div>
        </div>
      </main>
    </div>
  </body>,
  "container": <div>
    <main
      class="sc-iqAclL dFOpGW"
      data-testid="app-container"
    >
      <div
        class="sc-crzoAE eQQNpY"
        id="app-content"
      >
        <div
          class="sc-kEqXSa bYPdLh"
          id="app-header"
        >
          <div
            class="MuiFormControl-root MuiTextField-root sc-gtsrHT gePIXd"
          >
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
                id="search-input"
                placeholder="Search for your favourite song, artist or album"
                type="text"
                value=""
              />
              <div
                class="MuiInputAdornment-root MuiInputAdornment-positionStart"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M15.5 14h-.79l-.28-.27c1.2-1.4 1.82-3.31 1.48-5.34-.47-2.78-2.79-5-5.59-5.34-4.23-.52-7.79 3.04-7.27 7.27.34 2.8 2.56 5.12 5.34 5.59 2.03.34 3.94-.28 5.34-1.48l.27.28v.79l4.25 4.25c.41.41 1.08.41 1.49 0 .41-.41.41-1.08 0-1.49L15.5 14zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                  />
                </svg>
              </div>
              <fieldset
                aria-hidden="true"
                class="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
                style="padding-left: 8px;"
              >
                <legend
                  class="PrivateNotchedOutline-legend-2"
                  style="width: 0.01px;"
                >
                  <span>
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="sc-dlnjwi ksErfN"
          id="display-results-container"
        >
          <div
            class="sc-hKFxyN iiWwtn"
            id="display-results-content"
          />
        </div>
      </div>
    </main>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
